design/20220720-per-certificate-owner-ref.md

```diff
@@ -0,0 +1,339 @@
+and introduce the new flag `--default-secret-cleanup-policy`. Its values are as follows:
+
+- When `--default-secret-cleanup-policy` is set to `Never`, the Certificate resources
+  that don't have the `cleanupPolicy` field set will have their associated Secret
+  resources updated (i.e., the owner reference gets removed) on the next issuance of
+  the Certificate.
+- When `--default-secret-cleanup-policy` is set to `OnDelete`, the Certificate resources
+  that don't have the `cleanupPolicy` field set will have their associated Secret
+  resources updated (i.e., the owner reference gets added) on the next issuance of
+  the Certificate.
+  
+The effect of changing `--default-secret-cleanup-policy` from `Never` to `OnDelete`
+or from `OnDelete` to `Never` is not immediate: the change requires a re-issuance
+of the Certificate resources.
+
+The default value for `--default-secret-cleanup-policy` is `Never`.
+
+When changing the flag from `Never` to `OnDelete`, the existing Certificate resources
+that don't have `cleanupPolicy` set are immediately affected, meaning that their
+associated Secrets will gain a new owner reference. When changing the flag from
+`OnDelete` to `Never`, the Secrets associated to Certificates that have no `cleanupPolicy`
+set will see their owner reference immediately removed.
+
+The reason we decided to deprecate `--enable-certificate-owner-ref` is because this
+flat behaves differently to how the new `cleanupPolicy` behaves:
+
+- When `--enable-certificate-owner-ref` is not passed (or is set to false), the existing
+  Secret resources that have an owner reference are not changed even after a re-issuance.
+  With `--default-secret-cleanup-policy` and given that `cleanupPolicy` is not set, the
+  behavior is slightly different: unlike with the old flag, the existing Secret resources
+  will have their owner references removed.
+- When `--enable-certificate-owner-ref` is set to true, the behavior is the same as
+  when `--default-secret-cleanup-policy` is set to `OnDelete` and `cleanupPolicy` is not
+  set.  
+
+The deprecated flag `--enable-certificate-owner-ref` keeps precendence over the new flag
+in order to keep backwards compatibility.
+
+When upgrading to the new flag, users can refer to the following table:
+
+| If... | then they should replace it with... |
+|-----|-----|
+| `--enable-certificate-owner-ref` not passed to the controller | No change needed |
+| `--enable-certificate-owner-ref=false` | Replace with `--default-secret-cleanup-policy=Never` |
+| `--enable-certificate-owner-ref=true` | Replace with `--default-secret-cleanup-policy=OnDelete` |
+
+## Use-cases
+
+[Flant](https://flant.com) manages certificates for users, and has hit a Kubernetes apiserver limitation where too many leftover Secret resources were slowing the apiserver down. This issue has happened because Certificate resources are created using auto-generated names, and Certificate resources are often deleted shortly after being created.
```

Comment:
@maelvls We have clients with thousands of certificates in dev-namespaces, it is reasonable to remove secrets along with certificates by enabling --enable-certificate-owner-ref=true. On the other hand, there are certificates with secrets in the same cluster which cannot be removed, because restarting the ingress associated with these secrets violates the SLA. For these secrets, it makes sense to disable --enable-certificate-owner-ref=false.
In the current cert-manager implementation, you cannot do this.

Reply:
Thanks. I read https://github.com/deckhouse/deckhouse/pull/1601 and I now have a better understanding.

I also came across https://github.com/cert-manager/cert-manager/issues/3884. The author of this issue wants their Vault-issued certificates to add the owner reference on Secrets, but they don't want their Let's Encrypt-issued certificate to add the owner reference. I asked them what was their use-case.